### PR TITLE
[WIP] task/issue-275 Create Extensible Menus Query

### LINKED ISSUE #275: add support for dynamic imports
## Type of Change
- [ ] New Feature Request
- [ ] Documentation / Website
- [ ] Improvement / Suggestion
- [ ] Bug
- [x] Other (please clarify below)

## Summary
Currently the shelf uses `require` to load JSON files.
```javascript
setupShelf() {
  // based on path, display selected list
  const url = window.location.pathname;
  let list = [];

  if (url.indexOf('/about') >= 0) {
    list = require('../components/shelf/about.json');
  } else if (url.indexOf('/docs') >= 0) {
    list = require('../components/shelf/documentation-list.json');
  } else if (url.indexOf('/getting-started') >= 0) {
    list = require('../components/shelf/getting-started-list.json');
  } else if (url.indexOf('/plugins') >= 0) {
    list = require('../components/shelf/plugins.json');
  }

  this.shelfList = list;
}
```

## Details
This is only permitted because **webpack** allows it, but browser code should only be using browser APIs, like `import` / `export`.

The project should be made to support `import()` syntax and remove any usage of NodeJS specific code, like `require` in the shelf.

### PR DESCRIPTION
**Pull request by hutchgrant**
## Related Issue
Resolves #275

## Summary of Changes

1. Begins to address (#271) which requires deterministic queries through a single menu based query
2. Removes need for shelfList to use json file
3. Uses [markdown-toc](https://github.com/jonschlinkert/markdown-toc) to generate a table of contents from every page flagged with front-matter variable `linkheadings: true`
4. Added markdown front-matter globals for `menu, index, linkheadings` which are added to our graph via the graph lifecycle
5. if `linkheadings` is flagged true, during graph lifecycle(as markdown file is already being read) we will parse the file for all headings, add it to our graph for that specific page, into an array we will use for our shelf later.
6. the `menu` front-matter is used to specifically define a menu to use with any given markdown file e.g. `navigation` or `side`
7. I've modified our graphql  type defs to add `link` and `menu` 
8. I've added filtering to the query to only display listing for specific menus. e.g. `side` or `navigation`
9. I've modified our resolver to instead look through graph for any pages with the corresponding menu name
10. as per 9, I've also made the resolver check if a file has been flagged `listheadings` if so, it will loop through the `tableOfContents` variable for that page, and create a list of children for that specific link.

## TODO
1. shelf is not rendering the results correctly at the moment, this is important and is the next step, but the data however is being received correctly for all menus
2. fix broken collapse/render that was dependent on previous json file

## Note
At the moment this WIP. Data is received correctly, for all menus, only need to render it.

### CODE REVIEW

**Comment by hutchgrant** [comment]
I thought we were removing need for the menu json file altogether. This way it's done on a page by page basis determines what page should be in what menu, and if necessary, it will add all headings as a table of contents within the graph for the page.

File: www/components/shelf/shelf.js [hunk 1]
```diff
@@ -22,29 +22,24 @@ class Shelf extends LitElement {
...
-  }
-
-  async setupShelf(page) {
-    console.log('setupShelf for page =>', page);
-
-    if (page && page !== '' && page !== '/') {
-      // TODO remove require call - #275
-      this.shelfList = require(`./${page}.json`);
-
-      // TODO not actually integrated, still using .json files - #271
```
Comment (hutchgrant) [inline www/components/shelf/shelf.js]: This line is what I'm referring to when I resolved 271, using my method.  You knew we were going to use a query here.

**Comment by thescientist13** [comment]
Well the details of #275 are pretty clear I think, in that specifically we shouldn't be using NodeJS APIs (`require`) in browser code, where we should be using Browser APIs (`import()`).
> This is only permitted because webpack allows it, but browser code should only be using browser APIs, like import / export

I think per something I saw in #265 , this would require additional Babel configuration.  I can rename the issue if it helps clarify the intent but that was the outcome I was looking for from this issue.

For this PR then and as with most things, I still recommend that it all new ideas / enhancements goes through an issue workflow first ideally, to allow for technical design review / decision and share code snippets and APIs.  Given the amount of moving parts here, I think this warrants it and then the issue links and branch names can be reflective of the intended scope / issue.

**Comment by hutchgrant** [comment]
You specifically put a query, because the idea is to query our menu through graphql not import or require. There's no need to import or require that json file because its redundant with a working grpahql query.  That's the point I think you're missing here. Read the rest of your code for the component.

**Comment by thescientist13** [comment]
I understand the point, I wrote the issue.  :)

like I said, I'll just rename the issue **because** the main point was supporting `import()`, regardless of the use case.  You make it seem like no one could ever need to use it, ever which is shortsighted.  Also because `require` never should have been used in the first place.

My opinion on issue preceding PRs still stands and would / should apply here though, it is the way.

**Comment by hutchgrant** [comment]
closing this issue and i'll resubmit it

**Pull request closed without merging**